apps/api/src/app/subscribers/e2e/bulk-create-subscribers.e2e.ts

```diff
@@ -0,0 +1,149 @@
+            subscriberId: 'test2',
+            firstName: 'sub2',
+            email: 'sub2@test.co',
+          },
+          { subscriberId: subscriber.subscriberId, firstName: 'update name' },
+          { subscriberId: 'test2', firstName: 'update name' },
+        ],
+      },
+      {
+        headers: {
+          authorization: `ApiKey ${session.apiKey}`,
+        },
+      }
+    );
+
+    expect(body.data).to.be.ok;
+    const response = body.data;
+    const { updated, updatedCount, created, createdCount, failed } = response;
+
+    expect(updatedCount).to.equal(2);
+    expect(updated?.length).to.equal(2);
+    expect(updated[0].subscriberId).to.equal(subscriber.subscriberId);
+    expect(updated[1].subscriberId).to.equal('test2');
+
+    expect(createdCount).to.equal(2);
+    expect(created?.length).to.equal(2);
+    expect(created[0].subscriberId).to.equal('test1');
+    expect(created[1].subscriberId).to.equal('test2');
+
+    expect(failed?.length).to.equal(0);
+  });
+
+  it('should create and update subscribers', async function () {
+    const { data: body } = await axiosInstance.post(
+      `${session.serverUrl}${BULK_API_ENDPOINT}`,
+      {
+        subscribers: [
+          {
+            subscriberId: 'sub1',
+            firstName: 'John',
+            lastName: 'Doe',
+            email: 'john@doe.com',
+            phone: '+972523333333',
+            locale: 'en',
+            data: { test1: 'test value1', test2: 'test value2' },
+          },
+          {
+            subscriberId: 'test2',
+            firstName: 'sub2',
+            email: 'sub2@test.co',
+          },
+          {
+            subscriberId: 'test3',
+            firstName: 'sub3',
+            email: 'sub3@test.co',
+          },
+          { subscriberId: subscriber.subscriberId, firstName: 'update' },
+          {
+            subscriberId: 'test4',
+            firstName: 'sub4',
+            email: 'sub4@test.co',
+          },
+        ],
+      },
+      {
+        headers: {
+          authorization: `ApiKey ${session.apiKey}`,
+        },
+      }
+    );
+    expect(body.data).to.be.ok;
+    expect(body.data.updatedCount).to.equal(1);
+    expect(body.data.createdCount).to.equal(4);
+
+    const createdSubscriber = await subscriberRepository.findBySubscriberId(session.environment._id, 'sub1');
+    const updatedSubscriber = await subscriberRepository.findBySubscriberId(
+      session.environment._id,
+      subscriber.subscriberId
+    );
+
+    expect(updatedSubscriber?.firstName).to.equal('update');
+    expect(createdSubscriber?.firstName).to.equal('John');
+    expect(createdSubscriber?.email).to.equal('john@doe.com');
+    expect(createdSubscriber?.phone).to.equal('+972523333333');
+    expect(createdSubscriber?.locale).to.equal('en');
+    expect(createdSubscriber?.data?.test1).to.equal('test value1');
+  });
+
+  it('should throw an error when sending more than 500 subscribers', async function () {
+    const payload = {
```

Comment:
I can't think of a way to fake a failed update during the write executions... If the user is not providing enough information, it will be caught by the nested validator

Reply:
Ok, so assuming the nested validator is tested already to solve the other failure case we could provide proper uploads and fake the failure in the DAL service, for example, with Sinon. So in:
```
      bulkResponse = await this.bulkWrite(bulkWriteOps);
 ```
we would need to fake the response so for example from 3 subscribers, 2 go well and 1 fails. Therefore we can test the rest of the logic for the wrong case.
Anyway, if you need to move on other tasks let's not spend too much time on it. 👍🏻